content/blog/handling-react-remote-modules-with-babel.md

```diff
@@ -0,0 +1,267 @@
+            [
+              t.arrowFunctionExpression(
+                [],
+                t.callExpression(t.import(), [t.stringLiteral(componentPath)])
+              )
+            ]
+          )
+        )
+      ])
+    }
+  },
+};
+```
+
+And this actually worked, it was doing the transformation we wanted. But we weren't very comfortable on how difficult it was to read, and we found out a new way that looked much more readable:
+
+```js
+function moduleFederationReactImportAliases() {
+  return {
+    visitor: {
+      const componentPath = path.node.source.value;
+
+      if (path.node.source.value.startsWith('sharedcomps')) {
+        const componentPath = path.node.source.value;
+        const exportName = path.node.specifiers[0].local.name;
+
+        const transformed = babel.template.statement.ast`const ${exportName} = React.lazy(() => import('${componentPath}'))`;
+
+        path.insertAfter(transformed);
+      }
+    }
+  },
+};
+```
+
+This fits our goal, it was grabbing the correct import declarations, and replacing it with the new one that included the remote modules. 
+
+There was still a small problem. With the above logic, anything that was imported from this _remote components_ was being transformed into this `React.lazy` imports.
+
+What would happen if, together with the component `Button`, we imported a type, as in the following example:
+
+```js
+import Button, { ButtonProps } from "shared/components/Button";
+```
+
+As you're probably expecting, this would be transformed to:
+
+```js
+const Button = React.lazy(() => import("shared/components/Button"));
+```
+
+And we would loose the `ButtonProps` import, and that was wrong. We had to make a few changes so that we would only add the _dynamic import_ to the object that was being exported by default in the module. And this is how the final code looked like.
+
+```js
+function moduleFederationReactImportAliases(_, ...options) {
+  return {
+    visitor: {
+      ImportDeclaration: function (path, state) {
+        // console.log('opts', state)
+        if (path.node.source.value.includes('sharedcomps/Button')) {
+          const componentPath = path.node.source.value;
+          const defaultExportName = path.node.specifiers.find((spec) => spec.type === "ImportDefaultSpecifier").local.name;
+          const specWithoutDefault = path.node.specifiers.filter((spec) => spec.type !== "ImportDefaultSpecifier");
+          path.node.specifiers = specWithoutDefault;
+
+          const transformed = babel.template.statement.ast`const ${defaultExportName} = React.lazy(() => import('${componentPath}'))`;
+          path.insertAfter(transformed);
+
+          return
+        }
+      }
+    },
+  };
+}
+```
+
+The above code finds the `ImportDefaultSpecifier` and modifies the initial import declaration so that it keeps all except the default. Then it adds the code to do the dynamic import, together with `React.lazy`.
+## Conclusion
```

Comment:
Before the conclusion I would add the piece of code that allows us to decide on what components we want this to run on. It's open on a PR for dashboard-ui